File: Web/frontend/src/api/governance.ts
import { apiFetch } from "./client";

export type AssignedDesktop = {
  desktopAppId: string;
  nameLabel?: string;
  lastSeenAtUtc?: string;
  requiredApprovalsN: number;
  approvalsSoFar: number;
  status: "Pending" | "Active" | "Disabled";
  sessionStatus: "None" | "Pending" | "Unlocked" | "Expired" | "Cancelled";
  unlockedUntilUtc?: string;
  alreadyApproved?: boolean;
  remainingSeconds?: number;
};

export type ApprovalSummary = {
  sessionId: string;
  desktopAppId: string;
  status: AssignedDesktop["sessionStatus"];
  requiredApprovalsSnapshot: number;
  unlockedUntilUtc?: string;
  remainingSeconds?: number;
  approvals: { approverUserId: string; approvedAtUtc: string; approverEmail?: string }[];
};

export function getAssignedDesktops(token: string) {
  return apiFetch<AssignedDesktop[]>("/api/governance/desktops", { token });
}

export function approveDesktop(desktopAppId: string, token: string) {
  return apiFetch<ApprovalSummary>(`/api/governance/desktops/${desktopAppId}/approve`, {
    method: "POST",
    token,
  });
}

export function getDesktopHistory(desktopAppId: string, token: string) {
  return apiFetch<ApprovalSummary | null>(`/api/governance/desktops/${desktopAppId}/history`, { token });
}

File: Web/backend/app/schemas/governance.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel

from app.models.desktop import DesktopStatus
from app.models.session import SessionStatus


class AssignedDesktop(BaseModel):
    desktopAppId: str
    nameLabel: Optional[str] = None
    appType: Optional[str] = None
    lastSeenAtUtc: Optional[datetime] = None
    requiredApprovalsN: int
    approvalsSoFar: int
    status: DesktopStatus
    sessionStatus: SessionStatus
    unlockedUntilUtc: Optional[datetime] = None
    alreadyApproved: bool = False
    remainingSeconds: int = 0

    model_config = {"from_attributes": True}


class ApprovalItem(BaseModel):
    approverUserId: str
    approvedAtUtc: datetime
    approverEmail: Optional[str] = None

    model_config = {"from_attributes": True}


class ApprovalSummary(BaseModel):
    sessionId: str
    desktopAppId: str
    status: SessionStatus
    requiredApprovalsSnapshot: int
    unlockedUntilUtc: Optional[datetime] = None
    remainingSeconds: int = 0
    approvals: List[ApprovalItem]

    model_config = {"from_attributes": True}
